// ngx-stars/esm2020/lib/ngx-stars.module.mjs
import { NgModule } from '@angular/core';
import { NgxStarsComponent } from './ngx-stars.component';
import { CommonModule } from '@angular/common';
import * as i0 from "@angular/core";
export class NgxStarsModule {
}
NgxStarsModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: NgxStarsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxStarsModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "14.0.0", ngImport: i0, type: NgxStarsModule, declarations: [NgxStarsComponent], imports: [CommonModule], exports: [NgxStarsComponent] });
NgxStarsModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: NgxStarsModule, imports: [CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "14.0.0", ngImport: i0, type: NgxStarsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        NgxStarsComponent
                    ],
                    exports: [
                        NgxStarsComponent
                    ]
                }]
        }] });
